src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs

```diff
@@ -428,24 +428,101 @@ public virtual KeyedHashAlgorithm CreateKeyedHashAlgorithm(byte[] keyBytes, stri
             if (CustomCryptoProvider != null && CustomCryptoProvider.IsSupportedAlgorithm(algorithm, keyBytes))
             {
                 if (!(CustomCryptoProvider.Create(algorithm, keyBytes) is KeyedHashAlgorithm keyedHashAlgorithm))
-                    throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10647, LogHelper.MarkAsNonPII(algorithm), LogHelper.MarkAsNonPII(typeof(KeyedHashAlgorithm)))));
+                    throw LogHelper.LogExceptionMessage(
+                        new InvalidOperationException(
+                            LogHelper.FormatInvariant(
+                                LogMessages.IDX10647,
+                                LogHelper.MarkAsNonPII(algorithm),
+                                LogHelper.MarkAsNonPII(typeof(KeyedHashAlgorithm)))));
 
                 return keyedHashAlgorithm;
             }
 
+            // In the case of Aes128CbcHmacSha256, Aes192CbcHmacSha384, Aes256CbcHmacSha512 which are Authenticated Encryption algorithms
+            // SymmetricSignatureProvider will get passed a key with 1/2 the minimum keysize expected size for the HashAlgorithm. 16 bytes for SHA256, instead of 32 bytes.
+            // see: https://datatracker.ietf.org/doc/html/rfc7518#section-5.2.2.1
             switch (algorithm)
             {
+                case SecurityAlgorithms.Aes128CbcHmacSha256:
+                {
+                        if (keyBytes.Length < 16)
+                            throw LogHelper.LogExceptionMessage(
+                                new InvalidOperationException(
+                                    LogHelper.FormatInvariant(LogMessages.IDX10720,
+                                    LogHelper.MarkAsNonPII(algorithm),
+                                    LogHelper.MarkAsNonPII("128"),
+                                    LogHelper.MarkAsNonPII(keyBytes.Length * 8))));
```

Comment:
nit: Create something like:

```
private void ThrowIfKeysizeTooSmall(string alg, int keyBytesLength, int minKeySize)
{
...
}
```

to avoid code duplication and prevent typos if this gets extended.

Reply:
ThrowIfKeySizeTooSmall(keyBytes.Length, 64, algorithm);

        private static void ThrowIfKeySizeTooSmall(int keyLengthInBytes, int minSizeInBytes, string algorithm)
        {
            if (keyLengthInBytes < minSizeInBytes)
            {
                throw LogHelper.LogExceptionMessage(
                    new InvalidOperationException(
                        LogHelper.FormatInvariant(LogMessages.IDX10720,
                        LogHelper.MarkAsNonPII(algorithm),
                        LogHelper.MarkAsNonPII(minSizeInBytes * 8),
                        LogHelper.MarkAsNonPII(keyLengthInBytes * 8))));
            }
        }